internal/llrp/behavior_test.go

```diff
@@ -0,0 +1,723 @@
+//
+// Copyright (C) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package llrp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"math"
+	"reflect"
+	"testing"
+	"testing/quick"
+)
+
+// testROSpecProperties is a helper function
+// that validates various properties on an ROSpec
+// that should always be true (of an ROSpec to be sent to a Reader),
+// regardless of how that ROSpec is generated.
+func testROSpecProperties(t *testing.T, spec *ROSpec) {
+	t.Helper()
+
+	// These assume we're creating a new spec
+
+	if spec.ROSpecID == 0 {
+		t.Error("spec ID should not be 0")
+	}
+
+	if spec.ROSpecCurrentState != ROSpecStateDisabled {
+		t.Error("spec should be created in the Disabled state")
+	}
+
+	// We're currently controlling reporting at the ReaderConfig level,
+	// so we don't want to override it in the ROSpec itself.
+	if spec.ROReportSpec != nil {
+		t.Error("spec should not use an ROReportSpec")
+	}
+
+	if len(spec.AISpecs) == 0 {
+		t.Error("spec should have at least one AI Spec")
+	}
+}
+
+func TestImpinjEnableBool16(t *testing.T) {
+	if err := quick.Check(func(subtype uint32) bool {
+		data := impinjEnableBool16(subtype)
+		if len(data) != int(binary.BigEndian.Uint16(data[2:])) {
+			return false
+		}
+
+		if binary.BigEndian.Uint16(data) != 1023 {
+			return false
+		}
+
+		if binary.BigEndian.Uint32(data[4:]) != uint32(PENImpinj) {
+			return false
+		}
+
+		if binary.BigEndian.Uint32(data[8:]) != subtype {
+			return false
+		}
+
+		if binary.BigEndian.Uint16(data[12:]) != 1 {
+			return false
+		}
+
+		return true
+	}, nil); err != nil {
+		t.Error(err)
+	}
+}
+
+var fccFreqs = []Kilohertz{
+	902750, 903250, 903750, 904250, 904750, 905250, 905750,
+	906250, 906750, 907250, 907750, 908250, 908750, 909250,
+	909750, 910250, 910750, 911250, 911750, 912250, 912750,
+	913250, 913750, 914250, 914750, 915250, 915750, 916250,
+	916750, 917250, 917750, 918250, 918750, 919250, 919750,
+	920250, 920750, 921250, 921750, 922250, 922750, 923250,
+	923750, 924250, 924750, 925250, 925750, 926250, 926750,
+	927250,
+}
+
+// newImpinjCaps returns capabilities matching an Impinj Reader in an FCC region.
+func newImpinjCaps() *GetReaderCapabilitiesResponse {
+	powerTable := make([]TransmitPowerLevelTableEntry, 81)
+	for i := range powerTable {
+		powerTable[i] = TransmitPowerLevelTableEntry{
+			Index:              uint16(i + 1),
+			TransmitPowerValue: MillibelMilliwatt(i*25 + 1000),
+		}
+	}
+
+	receiveTable := make([]ReceiveSensitivityTableEntry, 42)
+	for i := range receiveTable {
+		receiveTable[i] = ReceiveSensitivityTableEntry{
+			Index:              uint16(i + 1),
+			ReceiveSensitivity: uint16(i + 9),
+		}
+	}
+	receiveTable[0].ReceiveSensitivity = 0
+
+	if len(fccFreqs) != 50 {
```

Comment:
good example of where using `require` is preferred.
```
require.Equal(t, 50, len(fccFreqs), "This code is based on the assumption that the FCC Frequency list has 50 entries.")
```

Yes, you must pass `t` to this function.